eclair-core/src/main/scala/fr/acinq/eclair/wire/internal/channel/version4/ChannelCodecs4.scala

```diff
@@ -452,6 +453,46 @@ private[channel] object ChannelCodecs4 {
         ("claimHtlcDelayedPenaltyTxs" | listOfN(uint16, claimHtlcDelayedOutputPenaltyTxCodec)) ::
         ("spent" | spentMapCodec)).as[RevokedCommitPublished]
 
+    // We don't bother removing the duplication across HTLCs: this is a short-lived state during which the channel
+    // cannot be used for payments.
```

Comment:
Have I been too lazy here? Should we go the extra mile and remove htlc duplication?

Reply:
I don't think we need the htlc deduplication, there won't be any in the initial opening or splice cases. But there is at least some factoring to do on `Codec[CommitmentSpec]`, `Codec[LocalCommit]`, etc.

Reply:
I refactored the common parts in https://github.com/ACINQ/eclair/pull/2614/commits/bf56a6166a482d35e49beb3f0cdec06da9e7b590

Reply:
Nit:

```diff
diff --git a/eclair-core/src/main/scala/fr/acinq/eclair/wire/internal/channel/version4/ChannelCodecs4.scala b/eclair-core/src/main/scala/fr/acinq/eclair/wire/internal/channel/version4/ChannelCodecs4.scala
index 59c7d4cfd..c8f79f322 100644
--- a/eclair-core/src/main/scala/fr/acinq/eclair/wire/internal/channel/version4/ChannelCodecs4.scala
+++ b/eclair-core/src/main/scala/fr/acinq/eclair/wire/internal/channel/version4/ChannelCodecs4.scala
@@ -89,13 +89,18 @@ private[channel] object ChannelCodecs4 {
       .typecase(true, minimalHtlcCodec(htlcs.collect(DirectedHtlc.incoming)).as[IncomingHtlc])
       .typecase(false, minimalHtlcCodec(htlcs.collect(DirectedHtlc.outgoing)).as[OutgoingHtlc])

-    /** HTLCs are stored separately to avoid duplicating data. */
-    def commitmentSpecCodec(htlcs: Set[DirectedHtlc]): Codec[CommitmentSpec] = (
-      ("htlcs" | setCodec(minimalDirectedHtlcCodec(htlcs))) ::
+    private def baseCommitmentSpecCodec(directedHtlcCodec: Codec[DirectedHtlc]): Codec[CommitmentSpec] = (
+      ("htlcs" | setCodec(directedHtlcCodec)) ::
         ("feeratePerKw" | feeratePerKw) ::
         ("toLocal" | millisatoshi) ::
         ("toRemote" | millisatoshi)).as[CommitmentSpec]

+    /** HTLCs are stored separately to avoid duplicating data. */
+    def minimalCommitmentSpecCodec(htlcs: Set[DirectedHtlc]): Codec[CommitmentSpec] = baseCommitmentSpecCodec(minimalDirectedHtlcCodec(htlcs))
+
+    /** HTLCs are stored in full, the codec is stateless but creates duplication between local/remote commitment, and across commitments */
+    def commitmentSpecCodec: Codec[CommitmentSpec] = baseCommitmentSpecCodec(htlcCodec)
+
     val outPointCodec: Codec[OutPoint] = lengthDelimited(bytes.xmap(d => OutPoint.read(d.toArray), d => OutPoint.write(d)))

     val txOutCodec: Codec[TxOut] = lengthDelimited(bytes.xmap(d => TxOut.read(d.toArray), d => TxOut.write(d)))
@@ -350,9 +355,9 @@ private[channel] object ChannelCodecs4 {
     private def commitmentCodec(htlcs: Set[DirectedHtlc]): Codec[Commitment] = (
       ("fundingTxStatus" | fundingTxStatusCodec) ::
         ("remoteFundingStatus" | remoteFundingStatusCodec) ::
-        ("localCommit" | localCommitCodec(commitmentSpecCodec(htlcs))) ::
-        ("remoteCommit" | remoteCommitCodec(commitmentSpecCodec(htlcs.map(_.opposite)))) ::
-        ("nextRemoteCommit_opt" | optional(bool8, nextRemoteCommitCodec(commitmentSpecCodec(htlcs.map(_.opposite)))))).as[Commitment]
+        ("localCommit" | localCommitCodec(minimalCommitmentSpecCodec(htlcs))) ::
+        ("remoteCommit" | remoteCommitCodec(minimalCommitmentSpecCodec(htlcs.map(_.opposite)))) ::
+        ("nextRemoteCommit_opt" | optional(bool8, nextRemoteCommitCodec(minimalCommitmentSpecCodec(htlcs.map(_.opposite)))))).as[Commitment]

     /**
      * When multiple commitments are active, htlcs are shared between all of these commitments.
@@ -451,12 +456,6 @@ private[channel] object ChannelCodecs4 {
     // We don't bother removing the duplication across HTLCs: this is a short-lived state during which the channel
     // cannot be used for payments.
     private val interactiveTxWaitingForSigsCodec: Codec[InteractiveTxSigningSession.WaitingForSigs] = {
-      val commitmentSpecCodec: Codec[CommitmentSpec] = (
-        ("htlcs" | setCodec(htlcCodec)) ::
-          ("feeratePerKw" | feeratePerKw) ::
-          ("toLocal" | millisatoshi) ::
-          ("toRemote" | millisatoshi)).as[CommitmentSpec]
-
       val unsignedLocalCommitCodec: Codec[UnsignedLocalCommit] = (
         ("index" | uint64overflow) ::
           ("spec" | commitmentSpecCodec) ::

```

Reply:
Done in https://github.com/ACINQ/eclair/pull/2614/commits/a05436886eb716a0baa20ac7323dd6792635f5b4